firstvoices/backend/serializers/story_serializers.py

```diff
@@ -125,6 +128,60 @@ class Meta(SiteContentLinkedTitleSerializer.Meta):
         )
 
 
+class StoryDetailUpdateSerializer(StorySerializer):
+    pages = serializers.PrimaryKeyRelatedField(
+        queryset=StoryPage.objects.all(),
+        allow_null=True,
+        many=True,
+        validators=[SameSite()],
+    )
+
+    def to_representation(self, instance):
+        data = StorySerializer(instance=instance, context=self.context).data
+        return data
+
+    def update(self, instance, validated_data):
+        if "pages" in validated_data:
+            with transaction.atomic():
+                updated_pages = validated_data["pages"]
+                existing_pages = instance.pages.all()
+                temp_story = Story.objects.create(site=instance.site)
+
+                # Ensure that all updated pages belong to the story
+                for page in updated_pages:
+                    if page not in existing_pages:
+                        raise serializers.ValidationError(
+                            f"Page with ID {page.id} does not belong to the story."
+                        )
+
+                # Get the intersection of the existing and updated pages
+                new_pages = [
+                    page for page in updated_pages if page in set(existing_pages)
+                ]
+
+                # Delete any existing pages that are not in the updated list to ensure no orphans
+                for page in [
```

Comment:
This is going to iterate twice through the list-- better to just iterate once and move the condition inside